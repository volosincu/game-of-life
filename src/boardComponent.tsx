import  React, { useContext, useEffect, useState } from 'react';
import { CellComponent } from './cellComponent';
import { store, CellInterface } from './store';


import './main.scss';

function createBoardElements(x: number, y: number, matrix: []): any[] {
    const rows: any[] = [];
    let k = 0;
    for (let i = 0; i < x; i++) {
        const cols = [];
        for (let j = 0; j < y; j++) {
            const cel: CellInterface = matrix[k];
            cols[j] = <CellComponent key={`cel-${cel.id}`} id={cel.id} value={cel.value}/>;
            k++;
        }
        rows[i] = (<div className="row" key={`row-${k}`}>{cols}</div>);
    }
    return rows;
}

export const BoardComponent = () => {

    const globalState = useContext(store);
    const { state, dispatch } = globalState;
    const [period, setPeriod] = useState(500);
    const {state: { grid: {x, y}, nextgen: {id}} } = globalState;

    const rows = createBoardElements(x, y, state.grid.matrix);

    useEffect(() => {
        let gen = 0;
        const interval = setInterval(() => {
            dispatch({ type: 'NEXT_GEN', payload: {nextgen: {id: ++gen}}});
        }, period);
        return () => clearInterval(interval);
    }, [period]);

    return (
        <div className="container">
            <div className="menu">
                <h3>Generatia: {id} <span> Perioada: {period} ms</span></h3>
                <p>Zoom: {x} x {y}</p>
                <button className="speed-btn" onClick={() => {
                    if(period>400){
                        setPeriod(period-300);
                    }
                }}>Speed Up</button>
                <button className="speed-btn" onClick={() => {
                    setPeriod(period+300);
                }}>Speed Down</button>
                <button className="speed-btn" onClick={() => {
                    setPeriod(Math.pow(10, 10));
                }}>Stop</button>
                <button className="speed-btn" onClick={() => {
                    setPeriod(2000);
                }}>Start</button>
                <button className="speed-btn" onClick={() => {
                    dispatch({ type: 'ZOOM', payload: {}});
                }}>Zoom</button>
                <input type="radio" name="game" onChange={() => {
                    dispatch({ type: 'RANDOM', payload: {}});
                }}/>
                <label htmlFor="gg">Random</label>
                <input type="radio" name="game" onChange={() => {
                    dispatch({ type: 'GLIDERS_GUN', payload: {}});
                }}/>
                <label htmlFor="flower">Gliders gun</label>
                <input type="radio" name="game" onChange={() => {
                    dispatch({ type: 'FLOWER', payload: {}});
                }}/>
                <label htmlFor="flower">Flower</label>
            </div>
            <div className="game">
                {rows}
            </div>
        </div>
    );
};
