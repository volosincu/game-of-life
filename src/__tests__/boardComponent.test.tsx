import React from "react";
import { shallow } from "enzyme";
import { BoardComponent } from "../boardComponent";

jest.mock('react', () => {
    const ActualReact = require.requireActual('react')
    return {
        ...ActualReact,
        useContext: () => ({ state: {
            grid: {
                x: 1,
                y: 1,
                matrix: [{id: 1, value: true}]
            }, nextgen: {id: 1}
            }, dispatch: jest.fn() }),
    }
})

/**
 @TODO test the behaviour
 const mockDispatch = jest.fn();

 jest.mock("react", ()=>{
   useContext: () => {
       return {
           dispatch: mockDispatch
       }
   }
});*/

describe("BoardComponent", ()=> {
    let wrapper: any;
    beforeEach(()=> {
        wrapper = shallow(<BoardComponent />);
    });

    it('renders correctly - toMatchSnapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe("Should have expected styling and layout", ()=> {
        it('should renders the element with \" container \" className', () => {
            expect(wrapper.find(".container")).toHaveLength(1);
        });

        it('should render the menu', () => {
            expect(wrapper.find(".menu")).toHaveLength(1);
        });
    });

    describe("Should render the grid from the state matrix", ()=> {
        /*it('@TODO', () => {
            expect(wrapper.find("")).toHaveLength(1);
        });*/
    });
});