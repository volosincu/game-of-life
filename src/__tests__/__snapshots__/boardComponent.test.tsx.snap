// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardComponent renders correctly - toMatchSnapshot 1`] = `ShallowWrapper {}`;
